### PR TITLE
Add support for approximate k-NN queries

### LINKED ISSUE #539: [FEATURE] Add knn query support
### Is your feature request related to a problem?

Currently it is not possible to construct a query using opensearch-java with a knn query within it using the `Query` type:

https://opensearch.org/docs/latest/search-plugins/knn/filter-search-knn/#boolean-filter-with-ann-search
https://opensearch.org/docs/latest/search-plugins/knn/filter-search-knn/#lucene-k-nn-filter-implementation

opensearch-java does not contain KnnQuery as a variant it supports and due to the [Query.Kind](https://github.com/opensearch-project/opensearch-java/blob/main/java-client/src/main/java/org/opensearch/client/opensearch/_types/query_dsl/Query.java#L69) being a enum and the [QueryVariant](https://github.com/opensearch-project/opensearch-java/blob/main/java-client/src/main/java/org/opensearch/client/opensearch/_types/query_dsl/QueryVariant.java#L44) interface requiring a kind being specified it cannot be implemented as a separate library.

### What solution would you like?

Given that opensearch-project/k-NN#524 adds precedent for adding KNN functionality into opensearch-java directly and has support for script scoring KNN integration, it seems reasonable that KNN query support be added directly into opensearch-java by adding a `KNN` variant to the `Query.Kind` enum and adding `KnnQuery` and `KnnQuery.Builder` classes to allow for the construction of these queries.

### What alternatives have you considered?

`Query.Kind` being an enum is restrictive in itself, preventing composability of client library support for non-bundled plugins. It could be that removing this restriction allows for the client to be built separately. But given opensearch-project/k-NN#524 was merged it seems like given if this restriction was removed, it still may be advantageous for consistency to add the KNN query support

### Do you have any additional context?

Its not clear whether this request is the exact same thing as https://github.com/opensearch-project/opensearch-java/issues/547

### PR DESCRIPTION
**Pull request by Xtansia**
### Description
Adds support for the approximate k-NN query type to pair with the support for the `knn_vector` type added in #524. Along with a couple bug fixes for that support.

### Issues Resolved
#539 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: java-client/src/main/java/org/opensearch/client/opensearch/indices/IndexSettings.java [hunk 1]
```diff
@@ -1066,7 +1066,7 @@ protected void serializeInternal(JsonGenerator generator, JsonpMapper mapper) {
 
 		}
 		if (this.knnAlgoParamEfSearch != null) {
-			generator.writeKey("knn.algo_param_ef_search");
+			generator.writeKey("knn.algo_param.ef_search");
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/opensearch/indices/IndexSettings.java:1069]: :+1:

File: java-client/src/main/java/org/opensearch/client/opensearch/indices/IndexSettings.java [hunk 2]
```diff
@@ -1066,7 +1066,7 @@ protected void serializeInternal(JsonGenerator generator, JsonpMapper mapper) {
 
 		}
 		if (this.knnAlgoParamEfSearch != null) {
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/opensearch/indices/IndexSettings.java:1068]: Not related to this change, but it seems like a bunch of KNN index settings are missed:

```
"settings" : {
        "index": {
            "knn": true,
            "knn.algo_param.m":90,
            "knn.algo_param.ef_construction":200,
            "knn.algo_param.ef_search":300
        }
    },
```

https://github.com/opensearch-project/k-NN/blob/main/RFC.md?plain=1

File: CHANGELOG.md [hunk 3]
```diff
@@ -28,8 +28,9 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 ## [Unreleased 2.x]
 
 ### Added
-- Add support for knn_vector field type ([#529](https://github.com/opensearch-project/opensearch-java/pull/524))
+- Add support for knn_vector field type ([#524](https://github.com/opensearch-project/opensearch-java/pull/524))
```
Comment (reta) [inline CHANGELOG.md:31]: :+1:

**Review by reta — DISMISSED** [review]

**Review by reta — APPROVED** [review]

**Review by dblock — APPROVED** [review]
Do we need changes to USER_GUIDE? Can be followed up later.

**Pull request merged**